File: 2024/01/01.rb
def filepath(filename, subfolder = '')
  File.join(*[File.expand_path(__dir__), subfolder, filename].compact)
end

path = filepath('input.txt')
# path = filepath('exinput.txt')
lines = File.read(path).split("\n")

# Part 1
values = lines.map { |line| line.split(/\s+/).map(&:to_i) }
list1, list2 = values.transpose.each(&:sort!)

distances = list1.zip(list2).map { |a, b| (b - a).abs }
p distances.sum

# Part 2
counts = list2.tally
sim_score = list1.sum { |x| x * (counts[x] || 0) }
p sim_score

File: 2024/01/01.js
const {
  readFileSync,
  toTally,
  join,
  resolve,
  transpose,
} = require("../../js_modules/helpers.js");

function filepath(filename, subfolder = "") {
  return join(resolve(__dirname), subfolder, filename);
}

let path = filepath("input.txt");
path = filepath("exinput.txt");
let lines = readFileSync(path, "utf-8").trim().split("\n");

// Part 1

let pairs = lines.map((line) => line.split(/\s+/).map(Number));
let [list1, list2] = transpose(pairs);
[list1, list2].forEach((arr) => arr.sort((a, b) => a - b));

let differences = list1.map((x, i) => Math.abs(x - list2[i]));
console.log(differences.sum());

// Part 2

let tally2 = toTally(list2);
let simScore = list1.sum((x) => x * (tally2.get(x) ?? 0));
console.log(simScore);
